### ISSUE TITLE
Skip LC0021 on page objects

### ISSUE THREAD

**Issue by jwikman** [StefanMaron/BusinessCentral.LinterCop#413]
I suggest that LC0021 do not trigger on Page objects. (Except API pages, if someone now happened to think that confirms on those is a good idea 😉).

I cannot really see any advantages of using `Confirm Management` when in a page object. Anyone else have any good arguments on why to implement that on pages?

The page is already a UI, so any `GuiAllowed()` arguments goes away.
You _can_ publish a UI page as SOAP WS, but that is not recommended  anyway.

The scenario I stumble upon now and then is a `Confirm()` in page actions, and that should never be executed if not in UI. So at least skip LC0021 for page actions. 🙂

**Comment by Arthurvdv** [comment]
While indeed the `Confirm Management` is most effective on other objects then pages, I'm sill a believer to also apply this there.

This makes the rule more consistent and, next to the UI pages as SOAP WS, the OnBeforeGuiAllowed event is helpful for some automating testing to suppress the dialog and as bonus could be used in a PTE where the customer doesn't want to have the confirm dialog shown.

Not ideal, but you could add an pragma on the beginning of each page object to exclude this rule on these objects.

```AL
#pragma warning disable LC0021 
page 50100 MyPage
{
```

**Comment by pri-kise** [comment]
Maybe having a setting for this?

I know that settings can make things complex, but most of the confirms, that we haven't replaced with `Confirm Management` are on pages.

**Comment by Arthurvdv** [comment]
> The `OnBeforeGuiAllowed` is an internal event, not useful for anyone but MS.

I thought it was an `IntegrationEvent`, now I see it's a `InternalEvent` 😳

You're absolutely right, without an event, the excursion along the `Confirm Management` indeed doesn't bring any added value. Thank you for pointing this out for me. And I agree with you on this, this rule indeed doesn't bring any value on page object types.

What is het best solution here?
A) Exclude _all_ objects of type Page?
B) Exclude objects of type Page, where the PageType <> API

In case of B) are there also other PageTypes, like `ReportProcessingOnly` and `XmlPort` where we need to raise this rule?


PS I'll also update the wiki page of this rule, there I've provided the wrong info it seems:

> * OnBeforeGuiAllowed event publisher for integration and/or testing purposes

**Comment by jwikman** [comment]
I vote for B)

I have no idea on when to use the PageTypes `ReportProcessingOnly`, `ReportPreview`, `XmlPort` etc. (They are not used in BaseApp at all). Maybe they are just used by runtime when running a Report or XmlPort...`

But they are all pages, so I don't know if ConfirmManagement would bring any value there either.

**Comment by Arthurvdv** [comment]
Good idea, and I'll will gladly follow your lead in this!

I'll try to have this in the prerelease next week.

**Comment by Arthurvdv** [comment]
The (pre)release version of [v0.30.10](https://github.com/StefanMaron/BusinessCentral.LinterCop/releases/tag/v0.30.10) will now exclude page objects from this rule, with the exception when the PageType is set to API.

**Comment by jwikman** [comment]
Thanks @Arthurvdv, it now seems to work just as I wanted it to!

At least it warns in codeunits and API pages, but not on other pages. 👍
